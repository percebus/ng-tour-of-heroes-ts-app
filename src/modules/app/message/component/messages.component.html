<div class="messages-component" *ngIf="messageService.messages.length">
  <h2>Messages</h2>
  <button type="button" class="clear" (click)="clear()">Clear messages</button>
  <div *ngFor="let message of messageService.messages">
    {{ message }}
  </div>
</div>
