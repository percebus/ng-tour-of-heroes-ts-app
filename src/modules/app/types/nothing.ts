export type Nullable<T> = T | null;

export type Optional<T> = T | undefined;

export type Tentative<T> = T | undefined | null;
