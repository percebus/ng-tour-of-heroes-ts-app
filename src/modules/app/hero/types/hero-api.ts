import { Observable } from 'rxjs';
import { Optional } from '../../types/nothing';
import { Hero } from './hero';

export interface HeroAPI {
  post(hero: Hero): Observable<Hero>;
  put(hero: Hero): Observable<any>; // FIXME
  delete(hero: Hero): Observable<Hero>;
  getById(id: number): Observable<Optional<Hero>>;
  getAll(): Observable<Hero[]>;
}
