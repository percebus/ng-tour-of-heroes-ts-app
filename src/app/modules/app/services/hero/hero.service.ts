import { Observable, of } from 'rxjs';
import { Injectable } from '@angular/core';
import { MessageService } from '../message/message.service';
import { Hero } from 'src/app/types/hero';
import { heroes } from 'src/data/mock/heroes';

@Injectable({
  providedIn: 'root',
})
export class HeroService {
  constructor(private messageService: MessageService) {}

  getAll(): Observable<Array<Hero>> {
    this.messageService.add('HeroService: fetched heroes');
    return of(heroes);
  }
}
