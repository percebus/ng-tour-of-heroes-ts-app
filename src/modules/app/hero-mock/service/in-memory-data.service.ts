import { Injectable } from '@angular/core';
import { InMemoryDbService } from 'angular-in-memory-web-api';
import { Hero } from '../../hero/types/hero';
import { heroes } from '../data/heroes';

@Injectable({
  providedIn: 'root',
})
export class InMemoryDataService implements InMemoryDbService {
  createDb() {
    return { heroes };
  }

  // Overrides the genId method to ensure that a hero always has an id.
  // If the heroes array is empty,
  // the method below returns the initial number (11).
  // if the heroes array is not empty, the method below returns the highest
  // hero id + 1.
  genId(heroes: Hero[]): number {
    const id =
      heroes.length > 0 ? Math.max(...heroes.map((oHero) => oHero.id)) : 10;

    return id + 1;
  }
}
