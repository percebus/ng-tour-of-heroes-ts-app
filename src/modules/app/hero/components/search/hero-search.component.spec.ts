import { HttpClient } from '@angular/common/http';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import {
  HttpClientTestingModule,
  HttpTestingController,
} from '@angular/common/http/testing';
import { HeroSearchComponent } from './hero-search.component';
import { HeroAPI } from '../../types/hero-api';
import { HeroService } from '../../service/hero.service';
import { MockHeroService } from '../../../hero-mock/services/mock/hero.mock-service';

describe('HeroSearchComponent', () => {
  let oHeroSearchComponent: HeroSearchComponent;
  let oComponentFixture: ComponentFixture<HeroSearchComponent>;
  let mockHeroService: HeroAPI;
  let refresh: jest.SpyInstance;

  describe('using MockHeroService', () => {
    beforeEach(() => {
      TestBed.configureTestingModule({
        imports: [HttpClientTestingModule],
        declarations: [HeroSearchComponent],
      })
        .overrideComponent(HeroSearchComponent, {
          set: {
            providers: [{ provide: HeroService, useClass: MockHeroService }],
          },
        })
        .compileComponents();

      oComponentFixture = TestBed.createComponent(HeroSearchComponent);
      mockHeroService =
        oComponentFixture.debugElement.injector.get(HeroService);

      oHeroSearchComponent = oComponentFixture.componentInstance;
      refresh = jest.spyOn(oHeroSearchComponent, 'refresh');

      oComponentFixture.detectChanges();
    });

    it('is instanceof HeroSearchComponent', () => {
      expect(oHeroSearchComponent).toBeInstanceOf(HeroSearchComponent);
    });

    describe('.refresh()', () => {
      it('got invoked OnInit', () => {
        expect(refresh).toHaveBeenCalled();
      });
    });

    describe('HTML', () => {
      describe('search input text box', () => {
        it('renders "Hero Search"', () => {
          const oHTMLElement = oComponentFixture.nativeElement //
            .querySelector('#search-component');

          expect(oHTMLElement.textContent).toBe('Hero Search');
        });
      });

      it('matches snapshot', () => {
        expect(oComponentFixture).toMatchSnapshot();
      });
    });
  });
});
