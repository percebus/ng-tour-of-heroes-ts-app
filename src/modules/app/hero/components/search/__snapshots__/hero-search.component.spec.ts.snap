// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroSearchComponent using MockHeroService HTML matches snapshot 1`] = `
<app-hero-search
  heroService={[Function MockHeroService]}
  heroes$={[Function AnonymousSubject]}
  refresh={[Function Function]}
  searchTerms={[Function Subject]}
>
  <div
    id="search-component"
  >
    <label
      for="search-box"
    >
      Hero Search
    </label>
    <input
      id="search-box"
    />
    <ul
      class="search-result"
    >
    </ul>
  </div>
</app-hero-search>
`;
